compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -4162,11 +4161,22 @@ object Types {
     // Boolean caches: 0 = uninitialized, -1 = false, 1 = true
     private var myStableHash: Byte = 0
     private var myGround: Byte = 0
+    private var myIsStable: Boolean = false
+    private var myIsStableComputed: Boolean = false
```

Comment:
- This adds two bytes to the footprint of AppliedType, this could be done in one byte like the two caches above.
- Are there benchmarks that show a need for this cache? And does that need outweight the increase in memory usage? (though it's possible the memory usage didn't actually increase if the fields replaced padding, you can check that using jol: https://github.com/lampepfl/dotty/discussions/14254)
- More importantly, is this cache actually correct without any form of cache invalidation? It seems unlikely: `isStable` calls `stripTypeVar` so depending on the current state of the constraints, it could give different results. This is why some caches check for `isProvisional`.
 
In general, we have to be extremely careful when we decide to cache something, cache invalidation bugs are among the hardest issues to debug.

Reply:
Yes, I think isProvisional should be checked, and caches are dicey. But it's not a solution either to add a common case to a hot method that is two orders of magnitude slower than the other, carefully optimized cases. I don't fully trust benchmarks here either. It's quite plausible that the very limited benchmarks we have don't test this particular case and will come back to cause some serious performance degradation elsewhere.

The easiest thing would be to just not do it. Don't try to push compile-time ops further since it's risky. Or put in under an experimental flag. Or, if we don't want to do that, bite the bullet and make sure the caches work.

Reply:
> This adds two bytes to the footprint of AppliedType, this could be done in one byte like the two caches above.

Indeed, I should follow the same implementation as the two other caches.

> Are there benchmarks that show a need for this cache?

I couldn't observe signifiant changes in the exiting synthetic benchmarks, so I added some micro benchmarks to test `isStable` in isolation. These confirm that not caching gives a run time linear in the size of the type, while caching gives a constant run time (see https://github.com/lampepfl/dotty/pull/15268#discussion_r885617617).

> And does that need outweight the increase in memory usage? (though it's possible the memory usage didn't actually increase if the fields replaced padding, you can check that using jol: https://github.com/lampepfl/dotty/discussions/14254)

Thanks for the pointer! I ran jol following your instructions. It reports 5 bytes of padding in the layout of `CachedAppliedType` before adding the `myIsStable` field, and 4 after.

Setup:
```bash
curl https://repo1.maven.org/maven2/org/openjdk/jol/jol-cli/0.16/jol-cli-0.16-full.jar -o jol-cli-0.16-full.jar
sbt -batch -error 'export scala3-bootstrapped / fullClasspath'
java -jar jol-cli-0.16-full.jar internals -cp [LAST LINE OF THE OUTPUT OF THE PREVIOUS COMMAND] 'dotty.tools.dotc.core.Types$CachedAppliedType'
```

Result before adding the `myIsStable` field:
```
dotty.tools.dotc.core.Types$CachedAppliedType object internals:
OFF  SZ                               TYPE DESCRIPTION                VALUE
  0   8                                    (object header: mark)      0x0000000000000001 (non-biasable; age: 0)
  8   4                                    (object header: class)     0xf8021e56
 12   1                            boolean Type.mightBeProvisional    true
 13   3                                    (alignment/padding gap)    
 16   4                                int CachedProxyType.myHash     0
 20   4                                int AppliedType.validSuper     0
 24   1                               byte AppliedType.myStableHash   0
 25   1                               byte AppliedType.myGround       0
 26   2                                    (alignment/padding gap)    
 28   4   dotty.tools.dotc.core.Types.Type AppliedType.tycon          null
 32   4    scala.collection.immutable.List AppliedType.args           null
 36   4   dotty.tools.dotc.core.Types.Type AppliedType.cachedSuper    null
Instance size: 40 bytes
Space losses: 5 bytes internal + 0 bytes external = 5 bytes total
```

Result after adding the `myIsStable` field:
```
dotty.tools.dotc.core.Types$CachedAppliedType object internals:
OFF  SZ                               TYPE DESCRIPTION                VALUE
  0   8                                    (object header: mark)      0x0000000000000001 (non-biasable; age: 0)
  8   4                                    (object header: class)     0xf8021e56
 12   1                            boolean Type.mightBeProvisional    true
 13   3                                    (alignment/padding gap)    
 16   4                                int CachedProxyType.myHash     0
 20   4                                int AppliedType.validSuper     0
 24   1                               byte AppliedType.myStableHash   0
 25   1                               byte AppliedType.myGround       0
 26   1                               byte AppliedType.myIsStable     0
 27   1                                    (alignment/padding gap)    
 28   4   dotty.tools.dotc.core.Types.Type AppliedType.tycon          null
 32   4    scala.collection.immutable.List AppliedType.args           null
 36   4   dotty.tools.dotc.core.Types.Type AppliedType.cachedSuper    null
Instance size: 40 bytes
Space losses: 4 bytes internal + 0 bytes external = 4 bytes total
```

> More importantly, is this cache actually correct without any form of cache invalidation?

No, we indeed need to use `isProvisional` as explained by Martin.

Is it correct not to use the period? When is it relevant for caching, and when is not?